### PR TITLE
[FS] Move code into try block to catch token disposed errors

### PR DESCRIPTION
**Pull request by quantumagi**
See https://stratisplatformuk.visualstudio.com/StratisBitcoinFullNode/_workitems/edit/4022.

### CODE REVIEW

File: src/Stratis.Bitcoin/P2P/Peer/NetworkPeerConnection.cs [hunk 1]
```diff
@@ -226,15 +226,19 @@ public async Task SendAsync(Payload payload, CancellationToken cancellation = de
...
+                if (cancellation != default(CancellationToken))
+                {
+                    cts = CancellationTokenSource.CreateLinkedTokenSource(cancellation, this.CancellationSource.Token);
+                    cancellation = cts.Token;
+                }
+                else
+                {
+                    cancellation = this.CancellationSource.Token;
+                }
+
```
Comment (MithrilMan) [inline src/Stratis.Bitcoin/P2P/Peer/NetworkPeerConnection.cs]: I think we should handle the specific exception we had ( `ObjectDisposedException` ) because in this scenario the reason for that is that the peer just disconnected, so we can just cancel the execution of the call with `throw new OperationCanceledException();` that's the equivalent of 
```
if (token.IsCancellationRequested)   
    throw new OperationCanceledException(token);  
```
basically just signals that the task has been canceled
So we should expect this behavior to be valid and could just handle it without flagging it as an error

basically I'm fine with moving it under the try, but I'd add another try catch around it just to handle ObjectDisposedException and in its catch, throw an `OperationCanceledException` that will be handled by the outer try/catch

Comment (MithrilMan) [reply src/Stratis.Bitcoin/P2P/Peer/NetworkPeerConnection.cs]: consider wrapping the code with another try/catch that explicitly catch the ObjectDisposedException
```
try    {
    if (cancellation != default(CancellationToken))
    {
        cts = CancellationTokenSource.CreateLinkedTokenSource(cancellation, this.CancellationSource.Token);
        cancellation = cts.Token;
    }
    else
    {
        cancellation = this.CancellationSource.Token;
    }
}
catch (ObjectDisposedException oEx){
    // peer already disconnected, no need to handle this as unexpected behavior, just cancel the task
    throw new OperationCancelledException();
}
```

Comment (MithrilMan) [reply src/Stratis.Bitcoin/P2P/Peer/NetworkPeerConnection.cs]: so basically it will throw the OperationCancelledException, and will be catched by the outer catch
```catch (OperationCanceledException)
            {
                this.peer.Disconnect("Connection to the peer has been terminated");
                this.logger.LogTrace("(-)[CANCELED_EXCEPTION]");
                throw;
            }```

Comment (quantumagi) [reply src/Stratis.Bitcoin/P2P/Peer/NetworkPeerConnection.cs]: @MithrilMan , Those are all good suggestions and there is nothing not to like about them. It's just a matter of taste whether they are really required at this time or whether we want to risk additional changes to what may be a P1 or whether we would rather address additional changes in a follow-up P2 pull request.

In a P2 we can perhaps look at why it is possible for us to call methods on an object that is being disposed OR why `Dispose` can be called on an object that is still executing. Maybe `NetworkPeerConnection` needs `lock` statements on its public accessible methods to prevent multiple methods from executing concurrently for the same instance of an object? We have enough concurrency going on to keep the CPU busy so there may not be much advantage in allowing that - instead many race conditions would be avoided.

Once this is fixed all that additional code may then be redundant and may have to be removed again.

**Comment by codingupastorm** [comment]
@zeptin @fassadlr @noescape00 Can you guys have a look at this? It's assigned to me but this isn't an area I know a ton about, I think you will be more helpful!

**Comment by codingupastorm** [comment]
This seems correct to me for cases where the connection dies almost instantly, but again tagging @fassadlr @noescape00

File: src/Stratis.Bitcoin/P2P/Peer/NetworkPeerConnection.cs (hunk 1 above)
Comment (fassadlr) [reply src/Stratis.Bitcoin/P2P/Peer/NetworkPeerConnection.cs]: Personally, I would catch the ObjectDisposedException and not "cover" it with something else.

We can handle it here and throw again so that the peer disconnects in the outer catch,

Comment (quantumagi) [reply src/Stratis.Bitcoin/P2P/Peer/NetworkPeerConnection.cs]: > consider wrapping the code with another try/catch that explicitly catch the ObjectDisposedException
> 
> ```
> try    {
>     if (cancellation != default(CancellationToken))
>     {
>         cts = CancellationTokenSource.CreateLinkedTokenSource(cancellation, this.CancellationSource.Token);
>         cancellation = cts.Token;
>     }
>     else
>     {
>         cancellation = this.CancellationSource.Token;
>     }
> }
> catch (ObjectDisposedException oEx){
>     // peer already disconnected, no need to handle this as unexpected behavior, just cancel the task
>     throw new OperationCancelledException();
> }
> ```

Done.

**Comment by fassadlr** [comment]
I have tested this PR by syncing test and main net to 100k blocks. Have also stopped and started the node on several occasions.
